pimd/pim_mlag.c

```diff
@@ -0,0 +1,346 @@
+			__func__,
+			zebra_mlag_lib_msgid_to_str(mlag_msg.msg_type, buf, 80),
+			mlag_msg.data_len, mlag_msg.msg_cnt);
+
+	switch (mlag_msg.msg_type) {
+	case MLAG_STATUS_UPDATE: {
+		struct mlag_status msg;
+
+		rc = zebra_mlag_lib_decode_mlag_status(s, &msg);
+		if (rc)
+			return (rc);
+		pim_mlag_process_mlagd_state_change(msg);
+	} break;
+	case MLAG_PEER_FRR_STATUS: {
+		struct mlag_frr_status msg;
+
+		rc = zebra_mlag_lib_decode_frr_status(s, &msg);
+		if (rc)
+			return (rc);
+		pim_mlag_process_peer_frr_state_change(msg);
+	} break;
+	case MLAG_VXLAN_UPDATE: {
+		struct mlag_vxlan msg;
+
+		rc = zebra_mlag_lib_decode_vxlan_update(s, &msg);
+		if (rc)
+			return rc;
+		pim_mlag_process_vxlan_update(&msg);
+	} break;
+	case MLAG_MROUTE_ADD: {
+		struct mlag_mroute_add msg;
+
+		rc = zebra_mlag_lib_decode_mroute_add(s, &msg);
+		if (rc)
+			return (rc);
+		pim_mlag_process_mroute_add(msg);
+	} break;
+	case MLAG_MROUTE_DEL: {
+		struct mlag_mroute_del msg;
+
+		rc = zebra_mlag_lib_decode_mroute_del(s, &msg);
+		if (rc)
+			return (rc);
+		pim_mlag_process_mroute_del(msg);
+	} break;
+	case MLAG_MROUTE_ADD_BULK: {
+		struct mlag_mroute_add msg;
+		int i = 0;
+
+		for (i = 0; i < mlag_msg.msg_cnt; i++) {
+
+			rc = zebra_mlag_lib_decode_mroute_add(s, &msg);
+			if (rc)
+				return (rc);
+			pim_mlag_process_mroute_add(msg);
+		}
+	} break;
+	case MLAG_MROUTE_DEL_BULK: {
+		struct mlag_mroute_del msg;
+		int i = 0;
+
+		for (i = 0; i < mlag_msg.msg_cnt; i++) {
+
+			rc = zebra_mlag_lib_decode_mroute_del(s, &msg);
+			if (rc)
+				return (rc);
+			pim_mlag_process_mroute_del(msg);
+		}
+	} break;
+	default:
+		break;
+	}
+	return 0;
+}
+
+/****************End of PIM Mesasge processing handler********************/
+
+int pim_zebra_mlag_process_up(void)
+{
+	if (PIM_DEBUG_MLAG)
+		zlog_debug("%s: Received Process-Up from Mlag", __func__);
+
+	return 0;
+}
+
+int pim_zebra_mlag_process_down(void)
+{
+	if (PIM_DEBUG_MLAG)
+		zlog_debug("%s: Received Process-Down from Mlag", __func__);
+
+	return 0;
+}
+
+static int pim_mlag_register_handler(struct thread *thread)
+{
+	uint32_t bit_mask = 0;
+
+	if (!zclient)
+		return -1;
+
+	SET_FLAG(bit_mask, (1 << MLAG_STATUS_UPDATE));
+	SET_FLAG(bit_mask, (1 << MLAG_MROUTE_ADD));
+	SET_FLAG(bit_mask, (1 << MLAG_MROUTE_DEL));
+	SET_FLAG(bit_mask, (1 << MLAG_DUMP));
+	SET_FLAG(bit_mask, (1 << MLAG_MROUTE_ADD_BULK));
+	SET_FLAG(bit_mask, (1 << MLAG_MROUTE_DEL_BULK));
+	SET_FLAG(bit_mask, (1 << MLAG_PIM_CFG_DUMP));
+	SET_FLAG(bit_mask, (1 << MLAG_VXLAN_UPDATE));
+	SET_FLAG(bit_mask, (1 << MLAG_PEER_FRR_STATUS));
+
+	if (PIM_DEBUG_MLAG)
+		zlog_debug("%s: Posting Client Register to MLAG mask:0x%x",
+			   __func__, bit_mask);
+
+	zclient_send_mlag_register(zclient, bit_mask);
+	return 0;
+}
+
+void pim_mlag_register(void)
+{
+	if (router->mlag_process_register)
+		return;
+
+	router->mlag_process_register = true;
+
+	thread_add_event(router->master, pim_mlag_register_handler, NULL, 0,
+			 NULL);
```

Comment:
Because of following reasons I am creating a new task.
1) It Involves IPC with Zebra, which can suspend the Thread.
2) In Future if it becomes Heavy processing api, it had performance Impact.

Because of these reasons I Created new Task.